include/etcpal/cpp/inet.h

```diff
@@ -0,0 +1,658 @@
+inline std::string SockAddr::ToString() const
+{
+  if (ETCPAL_IP_IS_V4(&addr_.ip))
+    return ip().ToString() + ':' + std::to_string(addr_.port);
+  else if (ETCPAL_IP_IS_V6(&addr_.ip))
+    return '[' + ip().ToString() + "]:" + std::to_string(addr_.port);
+  else
+    return std::string();
+}
+
+/// \brief Get the IP address from the SockAddr.
+inline IpAddr SockAddr::ip() const noexcept
+{
+  return addr_.ip;
+}
+
+/// \brief Get the port number from the SockAddr.
+inline uint16_t SockAddr::port() const noexcept
+{
+  return addr_.port;
+}
+
+/// \ingroup etcpal_cpp_inet
+/// \brief A wrapper for the EtcPal MAC address type.
+///
+/// Provides C++ syntactic sugar for working with MAC addresses.
+class MacAddr
+{
+public:
+  /// \brief Constructs a null MAC address by default.
+  MacAddr() = default;
+  // Note: this constructor is not explicit by design, to allow implicit conversions e.g.
+  //   etcpal::MacAddr sa = etcpal_c_function_that_returns_macaddr();
+  MacAddr(const EtcPalMacAddr& c_mac) noexcept;
+  MacAddr& operator=(const EtcPalMacAddr& c_mac) noexcept;
+  explicit MacAddr(const uint8_t* mac_data) noexcept;
+
+  const EtcPalMacAddr& get() const noexcept;
+  std::string ToString() const;
+  const uint8_t* data() const noexcept;
+
+  bool IsNull() const noexcept;
+
+  static MacAddr FromString(const std::string& mac_str) noexcept;
+
+private:
+  EtcPalMacAddr addr_{};
+};
+
+/// \brief Construct a MAC address copied from an instance of the C EtcPalMacAddr type.
+inline MacAddr::MacAddr(const EtcPalMacAddr& c_mac) noexcept : addr_(c_mac)
+{
+}
+
+/// \brief Assign an instance of the C EtcPalMacAddr type to an instance of this class.
+inline MacAddr& MacAddr::operator=(const EtcPalMacAddr& c_mac) noexcept
+{
+  addr_ = c_mac;
+  return *this;
+}
+
+/// \brief Construct a MAC address from its raw 6-byte array representation.
+/// \param[in] mac_data Pointer to the MAC data, an array of length #ETCPAL_MAC_BYTES.
+inline MacAddr::MacAddr(const uint8_t* mac_data) noexcept
+{
+  std::memcpy(addr_.data, mac_data, ETCPAL_MAC_BYTES);
+}
+
+/// \brief Get a reference to the underlying C type.
+inline const EtcPalMacAddr& MacAddr::get() const noexcept
+{
+  return addr_;
+}
+
+/// \brief Convert the MAC address to a string representation.
+///
+/// See etcpal_mac_to_string() for more information.
+inline std::string MacAddr::ToString() const
+{
+  char str_buf[ETCPAL_MAC_STRING_BYTES];
+  etcpal_mac_to_string(&addr_, str_buf, ETCPAL_MAC_STRING_BYTES);
+  return str_buf;
+}
+
+/// \brief Get the raw 6-byte array representation of a MAC address.
+/// \return Pointer to an array of length #ETCPAL_MAC_BYTES containing the address data.
+inline const uint8_t* MacAddr::data() const noexcept
+{
+  return addr_.data;
+}
+
+/// \brief Whether this MacAddr represents a null (all 0's) MAC address.
+inline bool MacAddr::IsNull() const noexcept
+{
+  return std::all_of(std::begin(addr_.data), std::end(addr_.data), [](uint8_t byte) { return byte == 0; });
```

Comment:
It strikes me that the C side should have a macro/check for this, and then IsNull would call it.